### PR TITLE
feat: CLI commands for secondary indexes

### LINKED ISSUE #1553: Sec. Indexes: CLI commands for creating and dropping indexes

### PR DESCRIPTION
**Pull request by islamaliev**
## Relevant issue(s)

Resolves #1553 

## Description

Adds following CLI commands:
- `client index create` that creates a new index on an existing collection
- `client index drop` that drop a named indexes on a collection
- `client index list` that lists all existing indexes in db or indexes for a specific collection

## Tasks

- [x] I made sure the code is well commented, particularly hard-to-understand areas.
- [x] I made sure the repository-held documentation is changed accordingly.
- [x] I made sure the pull request title adheres to the conventional commit style (the subset used in the project can be found in [tools/configs/chglog/config.yml](tools/configs/chglog/config.yml)).
- [x] I made sure to discuss its limitations such as threats to validity, vulnerability to mistake and misuse, robustness to invalidation of assumptions, resource requirements, ...

## How has this been tested?

Unit test, integration tests, manual tests

Specify the platform(s) on which this was tested:
- MacOS

### CODE REVIEW

**Review by fredcarle — CHANGES_REQUESTED** [review]
Good work Islam! I do have a few things that I'd like to see changed and discussed as you'll see bellow.

File: api/http/handlerfuncs.go [hunk 1]
```diff
@@ -227,6 +229,195 @@ func patchSchemaHandler(rw http.ResponseWriter, req *http.Request) {
...
+	type fieldResponse struct {
+		Name      string                `json:"name"`
+		Direction client.IndexDirection `json:"direction"`
+	}
+
+	type indexResponse struct {
+		Name   string          `json:"name"`
+		ID     uint32          `json:"id"`
+		Fields []fieldResponse `json:"fields"`
+	}
```
[the remark this answers is not in the document]
Comment (islamaliev) [reply api/http/handlerfuncs.go:279]: `client.IndexDescription` doesn't have `json` tags

[the remark this answers is not in the document]
Comment (fredcarle) [reply api/http/handlerfuncs.go:279]: You can add them :)

File: api/http/handlerfuncs_index.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+		handleErr(req.Context(), rw, err, http.StatusInternalServerError)
+		return
+	}
+
+	var data map[string]string
+	err = json.NewDecoder(req.Body).Decode(&data)
+	if err != nil {
+		handleErr(req.Context(), rw, err, http.StatusInternalServerError)
+		return
+	}
```
Comment (fredcarle) [inline api/http/handlerfuncs_index.go:33]: suggestion: You can replace this with 
```go
err := getJSON(req, &data)
if err != nil {
	handleErr(req.Context(), rw, err, http.StatusBadRequest)
	return
}
```

File: api/http/handlerfuncs_index.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
```
Comment (fredcarle) [inline api/http/handlerfuncs_index.go:80]: suggestion: Same as above.

File: api/http/handlerfuncs_index.go [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+		}
+		colIndexes := collectionIndexes{Collections: indexesPerCol}
+		sendJSON(
+			req.Context(),
+			rw,
+			struct {
+				Data collectionIndexes `json:"data"`
+			}{Data: colIndexes},
+			http.StatusOK,
+		)
```
Comment (fredcarle) [inline api/http/handlerfuncs_index.go:132]: suggestion: I think you can replace all this with 
```go
sendJSON(
	req.Context(),
	rw,
	simpleDataResponse("collections", indexesPerCol),
	http.StatusOK,
)
```

File: cli/index_create.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+				return NewErrFailedToSendRequest(err)
+			}
+
+			response, err := io.ReadAll(res.Body)
+			if err != nil {
+				return NewErrFailedToReadResponseBody(err)
+			}
+			if e := res.Body.Close(); e != nil {
+				return NewErrFailedToCloseResponseBody(err)
+			}
```
Comment (fredcarle) [inline cli/index_create.go:92]: todo: `e` should be changed to `err`.

File: cli/index_drop.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+				return NewErrFailedToSendRequest(err)
+			}
+
+			response, err := io.ReadAll(res.Body)
+			if err != nil {
+				return NewErrFailedToReadResponseBody(err)
+			}
+			if e := res.Body.Close(); e != nil {
+				return NewErrFailedToCloseResponseBody(err)
+			}
```
Comment (fredcarle) [inline cli/index_drop.go:88]: todo: e should be changed to err.

File: cli/index_list.go [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+				return NewErrFailedToSendRequest(err)
+			}
+
+			response, err := io.ReadAll(res.Body)
+			if err != nil {
+				return NewErrFailedToReadResponseBody(err)
+			}
+			if e := res.Body.Close(); e != nil {
+				return NewErrFailedToCloseResponseBody(err)
+			}
```
Comment (fredcarle) [inline cli/index_list.go:75]: todo: e should be changed to err.

File: cli/index_create_test.go [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+	_, hasErrors := r["errors"]
+	assert.True(t, hasErrors, "command should return error")
+}
+
+func TestIndexCreateCmd_IfNoErrors_ShouldReturnData(t *testing.T) {
+	cfg, close := startNode(t)
+	defer close()
+
+	addSchemaCmd := MakeSchemaAddCommand(cfg)
+	err := addSchemaCmd.RunE(addSchemaCmd, []string{`type User { Name: String }`})
```
Comment (fredcarle) [inline cli/index_create_test.go:110]: todo: We agreed to follow the convention of `Type` and `field` for our schemas. This should be changes to `type User { name: String }`.

File: cli/index_list_test.go [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+	err := indexCreateCmd.RunE(indexCreateCmd, nil)
+	require.ErrorIs(t, err, NewErrFailedToSendRequest(err))
+}
+
+func TestIndexListCmd_IfNoErrors_ShouldReturnData(t *testing.T) {
+	cfg, close := startNode(t)
+	defer close()
+
+	addSchemaCmd := MakeSchemaAddCommand(cfg)
+	err := addSchemaCmd.RunE(addSchemaCmd, []string{`type User { Name: String }`})
```
Comment (fredcarle) [inline cli/index_list_test.go:52]: todo: We agreed to follow the convention of `Type` and `field` for our schemas. This should be changes to `type User { name: String }`.

File: tests/integration/cli/client_index_create_test.go [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+// licenses/APL.txt.
+
+package clitest
+
+import (
+	"testing"
+)
+
+func createUserCollection(t *testing.T, conf DefraNodeConfig) {
+	createCollection(t, conf, `type User { Name: String }`)
```
Comment (fredcarle) [inline tests/integration/cli/client_index_create_test.go:18]: todo: We agreed to follow the convention of `Type` and `field` for our schemas. This should be changes to `type User { name: String }`.

File: tests/integration/cli/client_index_list_test.go [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+type indexResp struct {
+	Name   string           `json:"Name"`
+	Fields []indexFieldResp `json:"Fields"`
+}
+
+func TestIndexList_IfCollectionIsNotSpecified_ShouldReturnAllIndexes(t *testing.T) {
+	conf := NewDefraNodeDefaultConfig(t)
+	stopDefra := runDefraNode(t, conf)
+
+	createCollection(t, conf, `type User { Name: String }`)
```
Comment (fredcarle) [inline tests/integration/cli/client_index_list_test.go:33]: todo: We agreed to follow the convention of `Type` and `field` for our schemas. This should be changes to `type User { name: String }`.

File: tests/integration/cli/client_index_list_test.go [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+	createCollection(t, conf, `type Product { Name: String Price: Int }`)
```
Comment (fredcarle) [inline tests/integration/cli/client_index_list_test.go:34]: todo: We agreed to follow the convention of `Type` and `field` for our schemas. This should be changes to `type Product { name: String, price: Int }`.

**Review by fredcarle — CHANGES_REQUESTED** [review]
Good improvements Islam. I have a few more things before I approve and then we should be good to go. I'm sorry if some of the comments are repetitive. I used some copy pasting to save me time :)

File: cli/index_create.go [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+		Short: "Creates a secondary index on a collection's field(s)",
+		Long: `Creates a secondary index on a collection's field(s).
+		
+The --name flag is optional. If not provided, a name will be generated automatically.
+
+Example: create an index for 'Users' collection on 'Name' field:
+  defradb client index create --collection 'Users' --fields 'Name'
+
+Example: create a named index for 'Users' collection on 'Name' field:
+  defradb client index create --collection Users --fields Name --name UsersByName`,
```
Comment (shahzadlone) [inline cli/index_create.go:52]: todo: Consistent use of `'` in examples.

One example has:
`--collection Users --fields Name`
Other has:
`--collection 'Users' --fields 'Name'`

File: cli/index_create.go [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+				if err != nil {
+					return NewErrFailedToUnmarshalResponse(err)
+				}
+				if len(r.Errors) > 0 {
+					log.FeedbackError(cmd.Context(), "Failed to create index",
+						logging.NewKV("Errors", r.Errors))
+				} else {
+					log.FeedbackInfo(cmd.Context(), "Successfully created index",
+						logging.NewKV("Index", r.Data.Index))
+				}
```
Comment (shahzadlone) [inline cli/index_create.go:113]: todo: Seems like we want this `else` case tested.

Note: missing tests for the entire parent `else` block.

File: cli/index_create_test.go [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	indexCreateCmd := MakeIndexCreateCommand(cfg)
+	b := bytes.NewBufferString("")
+	indexCreateCmd.SetOut(b)
+
+	indexCreateCmd.SetArgs([]string{"--collection", "User",
+		"--fields", "name", "--name", "users_name_index"})
```
Comment (shahzadlone) [inline cli/index_create_test.go:120]: nitpick: Slight pref for:

```suggestion
	indexCreateCmd.SetArgs(
		[]string{"--collection", "User","--fields", "name", "--name", "users_name_index"},
	)
```

OR

```suggestion
	indexCreateCmd.SetArgs([]string{
		"--collection", "User",
		"--fields", "name",
		"--name", "users_name_index",	
		},
	)
```

File: cli/index_drop.go [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+				err = json.Unmarshal(response, &r)
+				if err != nil {
+					return NewErrFailedToUnmarshalResponse(err)
+				}
+				if len(r.Errors) > 0 {
+					log.FeedbackError(cmd.Context(), "Failed to drop index",
+						logging.NewKV("Errors", r.Errors))
+				} else {
+					log.FeedbackInfo(cmd.Context(), "Successfully dropped index",
+						logging.NewKV("Result", r.Data.Result))
```
Comment (shahzadlone) [inline cli/index_drop.go:108]: todo: similar comment to above, seems to be missing tests for this case.

File: cli/index_list.go [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+				}
+				if len(r.Errors) > 0 {
+					log.FeedbackError(cmd.Context(), "Failed to list index",
+						logging.NewKV("Errors", r.Errors))
+				} else if collectionArg != "" {
+					log.FeedbackInfo(cmd.Context(), "Fetched indexes for collection "+collectionArg,
+						logging.NewKV("Indexes", r.Data.Indexes))
+				} else {
+					log.FeedbackInfo(cmd.Context(), "Fetched all indexes",
+						logging.NewKV("Collections", r.Data.Collections))
```
Comment (shahzadlone) [inline cli/index_list.go:98]: todo: Similar to other comments, missed test case?

File: tests/integration/cli/client_index_drop_test.go [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+
+const userColIndexOnNameFieldName = "users_name_index"
+
+func createIndexOnName(t *testing.T, conf DefraNodeConfig) {
+	createIndexOnField(t, conf, "User", "name", userColIndexOnNameFieldName)
+}
+
+func createIndexOnField(t *testing.T, conf DefraNodeConfig, colName, fieldName, indexName string) {
+	runDefraCommand(t, conf, []string{"client", "index", "create",
+		"--collection", colName, "--fields", fieldName, "--name", indexName})
```
Comment (shahzadlone) [inline tests/integration/cli/client_index_drop_test.go:31]: nitpick: similar format pref as other nitpick.

File: tests/integration/cli/client_index_drop_test.go [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+
+func TestIndexDrop_IfNoNameArg_ShouldFail(t *testing.T) {
+	conf := NewDefraNodeDefaultConfig(t)
+	stopDefra := runDefraNode(t, conf)
+
+	createUserCollection(t, conf)
+	createIndexOnName(t, conf)
+
+	_, stderr := runDefraCommand(t, conf, []string{"client", "index", "drop",
+		"--collection", "User"})
```
Comment (shahzadlone) [inline tests/integration/cli/client_index_drop_test.go:42]: nitpick: similar format pref as other nitpick.

File: tests/integration/cli/client_index_drop_test.go [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+
+func TestIndexDrop_IfNoCollectionArg_ShouldFail(t *testing.T) {
+	conf := NewDefraNodeDefaultConfig(t)
+	stopDefra := runDefraNode(t, conf)
+
+	createUserCollection(t, conf)
+	createIndexOnName(t, conf)
+
+	_, stderr := runDefraCommand(t, conf, []string{"client", "index", "drop",
+		"--name", "users_name_index"})
```
Comment (shahzadlone) [inline tests/integration/cli/client_index_drop_test.go:56]: nitpick: similar format pref as other nitpick.

File: tests/integration/cli/client_index_drop_test.go [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+
+func TestIndexDrop_IfCollectionWithIndexExists_ShouldDropIndex(t *testing.T) {
+	conf := NewDefraNodeDefaultConfig(t)
+	stopDefra := runDefraNode(t, conf)
+
+	createUserCollection(t, conf)
+	createIndexOnName(t, conf)
+
+	stdout, _ := runDefraCommand(t, conf, []string{"client", "index", "drop",
+		"--collection", "User", "--name", "users_name_index"})
```
Comment (shahzadlone) [inline tests/integration/cli/client_index_drop_test.go:70]: nitpick: similar format pref as other nitpick.

File: tests/integration/cli/client_index_drop_test.go [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+	assertNotContainsSubstring(t, stdout, "errors")
+	assertNotContainsSubstring(t, nodeLog, "errors")
+}
+
+func TestIndexDrop_IfCollectionDoesNotExist_ShouldFail(t *testing.T) {
+	conf := NewDefraNodeDefaultConfig(t)
+	stopDefra := runDefraNode(t, conf)
+
+	stdout, _ := runDefraCommand(t, conf, []string{"client", "index", "drop",
+		"--collection", "User", "--name", "users_name_index"})
```
Comment (shahzadlone) [inline tests/integration/cli/client_index_drop_test.go:84]: nitpick: similar format pref as other nitpick.

File: tests/integration/cli/client_index_drop_test.go [hunk 23]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+func TestIndexDrop_IfInternalError_ShouldFail(t *testing.T) {
+	conf := NewDefraNodeDefaultConfig(t)
+	stopDefra := runDefraNode(t, conf)
+
+	createUserCollection(t, conf)
+
+	stdout, _ := runDefraCommand(t, conf, []string{"client", "index", "drop",
+		"--collection", "User", "--name", "users_name_index"})
```
Comment (shahzadlone) [inline tests/integration/cli/client_index_drop_test.go:97]: nitpick: similar format pref as other nitpick.

File: tests/integration/cli/client_index_list_test.go [hunk 24]
```diff
@@ -0,0 +1,93 @@
...
+
+	var resp struct {
+		Data struct {
+			Indexes []client.IndexDescription `json:"indexes"`
+		} `json:"data"`
+	}
+	err := json.Unmarshal([]byte(stdout[0]), &resp)
+	require.NoError(t, err)
+
+	//expectedResp := indexResp{Name: userColIndexOnNameFieldName, Fields: []indexFieldResp{{Name: "name"}}}
```
Comment (shahzadlone) [inline tests/integration/cli/client_index_list_test.go:75]: todo: Remove comment?

**Review by shahzadlone — APPROVED** [review]
Some nitpicks above formatting, :P might be considered personal pref though. One minor todo.

Approving as all else LGTM =)

**Review by fredcarle — APPROVED** [review]
LGTM. Note that I do agree with Shahzad that changing the formatting of the command flags would help readability.

**Pull request merged**